### ISSUE TITLE
Agents With Manual Proxy Settings

### ISSUE THREAD

**Issue by ianlovering** [EmpireProject/Empire#53]
Hi,

I've been testing this while trying to work through a proxy server. 

As a normal user the agent correctly picks up the proxy settings from Internet Explorer and calls back.

When trying to run the agent as System it fails as there are no proxy settings set for the System user. When setting the proxy details manually the stager uses them correctly and sets up the agent; however, the agent does not use the manual proxy settings and is unable to communicate with the server.

Basically I'm left with an agent that never updates and only responds to the 'info' command.

I had a very quick look through the code in agent.ps1 and it looks as if it is never passed the manual proxy settings.

This looks like a awesome tool that will be very useful.

**Comment by enigma0x3** [comment]
We will look into this.

Thanks

**Comment by enigma0x3** [comment]
@ianlovering, just as an update. We have confirmed that this issue exists and after looking, it is going to take a little bit of additional work to get it all implemented correctly. We have this addition in the horizon and will start working on it shortly. It has been a very busy last few weeks for all of us, so I apologize for the delay on this.

**Comment by enigma0x3** [comment]
Hey @ianlovering,
Just as an update, we haven't forgotten about this. To implement, we need to shuffle a few things architecture wise. We have this in the pipeline and will get to it soon!

Thanks for the patience :)

**Comment by stoked-security** [comment]
Hi, are you  working on solving this? I am currently using Empire in an environment where a Domain Admin has no proxy set, so I am experiencing similar issue as the original reporter.

I was looking into creating a sort of proxy similar to the redirector module, but instead of using the netsh portproxy functionality I want to create something that opens a TCP server on the local host and forwards the received data over one of the existing proxy aware agents. I am looking through the Empire source code, but I have some trouble finding the correct way to send data over one of the existing agent connections. Could you maybe point me in the right direction?

**Comment by xorrior** [comment]
@ianlovering working on this today. :)

**Comment by kellybenim** [comment]
Hey guys, how do I get configure proxy in stagger? What is the syntax for using proxy parameters and proxycreds? For example, how do i to configure the proxy socks5 127.0.0.0 on port 9050?

**Comment by xorrior** [comment]
@ianlovering and anyone else, pushed some code for this https://github.com/EmpireProject/Empire/commit/dcf7207fb60ccd9dfcc6116d13a39b0e3d20a0b7 . Can you test please?

**Comment by kellybenim** [comment]
Ok guys, but into this file ../Empire/lib/listeners/http.py, how should I manually configure the proxy socks5 127.0.0.1 on port 9050?

**Comment by xorrior** [comment]
@kellybenim proxy settings are set in stager options.

**Comment by xorrior** [comment]
@ianlovering manual proxy settings with credentials have been added.

**Comment by paleos** [comment]
Hi xorrior,

I;ve tested last DEV version and it seems taht there is still problem, that throug proxy are sent only first 3 http requests (reported in #115 https://github.com/EmpireProject/Empire/issues/115), next http request are sent directly :-(.

p.

**Comment by xorrior** [comment]
Is this with PowerShell or Python launchers?

**Comment by paleos** [comment]
Hi xorrios,

Powershell, windows/launcher_bat.

"@echo off start /b powershell -noP -sta -w 1 -enc <payload>"

p.

**Comment by xorrior** [comment]
Manual proxy settings will now be passed to the agent (PowerShell). Python agents will be addressed soon.

**Comment by paleos** [comment]
Hi xorrios,

I've just tested and in DEV branch it seems that it works. The problem with only first 3 http reqs through proxy ( #115) is solved.

P.S.: unfortunatelly without obfuscation the windows/launcher_bat (powershell) is catched by Symantec yet and with OBFUSCATE option I was not able to make it work :-( (I will make some other tests and maybe I will create a new issue). Anyway thanks for your work
